### DISCUSSION TITLE
Using `serde_with` to deserialize from string or struct on a vector

### DISCUSSION THREAD

**Discussion by colmanhumphrey** [jonasbb/serde_with#333] [Q&A] [answered]
Thanks for the crate @jonasbb!

Building [off this example](https://github.com/serde-rs/serde/issues/723#issuecomment-871016087), which is an extension  from [serde's docs](https://serde.rs/string-or-struct.html).

We'll assume we have the same struct `Build` as in the serde docs:

```rust
#[derive(Debug, Deserialize)]
struct Build {
    context: String,
    dockerfile: Option<String>,
    #[serde(default)]
    args: Map<String, String>,
}
```

Let's now assume we wanted to deserialize potentially many of them, e.g our target struct is:

```rust
#[derive(Deserialize, Debug)]
pub struct ManyBuilds {
    some_fields_with_overall_info: i32, // etc
    builds: Vec<Build>
}
```

@jonasbb's suggestion is to annotate `builds` like (do I have this right Jonas?):

```rust
#[derive(Deserialize, Debug)]
pub struct ManyBuilds {
    some_fields_with_overall_info: i32, // etc
    #[serde_as(as = "Vec<PickFirst<(_, DisplayFromStr)>>")]
    builds: Vec<Build>
}
```

This looks great, although I am currently getting the following error:

```rust
error: expected an inert attribute, found an attribute macro
  --> src/main.rs:90:5
   |
90 |     #[serde_as(as = "Vec<PickFirst<(_, DisplayFromStr)>>")]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

I'm guessing there's some very simple solution, but I don't yet see it. Thanks!

**Comment by jonasbb** [comment, marked as answer]
That is my mistake, I was too brief in the other message. To use `serde_as` you need to place the proc-macro attribute `#[serde_with::serde_as]` before the `derive`. This section of the [user guide](https://docs.rs/serde_with/1.9.4/serde_with/guide/serde_as/index.html) explains how to setup everything and what you can do with it.

```rust
#[serde_with::serde_as]
#[derive(Deserialize, Debug)]
pub struct ManyBuilds {
    some_fields_with_overall_info: i32, // etc
    #[serde_as(as = "Vec<PickFirst<(_, DisplayFromStr)>>")]
    builds: Vec<Build>
}
```

Side-Note: The associated `Err` type on the `FromStr` implementations needs to implement `Display` for this to work.

**Reply by colmanhumphrey** [reply]
Thanks Jonas - that works! That's not your mistake at all - I had indeed previously correctly annotated the struct when trying the package a couple days ago (after reading the guide - I didn't catch onto PickFirst on that read however!), and just must have been sloppy when trying this example out, my bad.

Makes sense about the error, not currently a big constraint for my needs. 

Thanks again for the prompt responses!
